dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import types
+import math
+import logging
+import numpy as np
+
+class _DaliBaseIterator(object):
+    def __init__(self,
+                 pipelines,
+                 size=-1,
+                 reader_name=None,
+                 auto_reset=False,
+                 fill_last_batch=True,
+                 last_batch_padded=False):
+
+        assert pipelines is not None, "Number of provided pipelines has to be at least 1"
+        if not isinstance(pipelines, list):
+            pipelines = [pipelines]
+        self._num_gpus = len(pipelines)
+        # frameworks expect from its data iterators to have batch_size field,
+        # so it is not possible to use _batch_size instead
+        self.batch_size = pipelines[0].batch_size
+        self._size = int(size)
+        self._auto_reset = auto_reset
+
+        self._fill_last_batch = fill_last_batch
+        self._last_batch_padded = last_batch_padded
+        assert self._size != 0, "Size cannot be 0"
+        assert self._size > 0 or (self._size < 0 and (len(pipelines) == 1 or reader_name)), "Negative size is supported only for a single pipeline"
+        assert not reader_name or (reader_name and self._size < 0), "When reader_name is provided, size should not be set"
+        if self._size < 0 and not reader_name:
+            self._auto_reset = False
+            self._fill_last_batch = False
+            self._last_batch_padded = False
+        self._pipes = pipelines
+        self._counter = 0
+
+        # Build all pipelines
+        for p in self._pipes:
+            with p._check_api_type_scope(types.PipelineAPIType.ITERATOR):
+                p.build()
+
+        self._reader_name = reader_name
+        if self._reader_name:
```

Comment:
Maybe put it in some separate function like `_extract_from_reader_and_validate`? 
Just a suggestion, but this seems like it is a bit independent, especially as this is in one big `if` that is almost longer than a screen.

Reply:
OTOH those arguments are overwritten here directly.

You can probably swap all those check in one generic function, especially if we have some way of accessing them by name.

Reply:
Done - as much as I think it was worth